### PR TITLE
optimize ps v3 restore

### PR DESCRIPTION
**Pull request by hehechen**
Signed-off-by: hehechen <awd123456sss@gmail.com>

### What problem does this PR solve?

Issue Number: ref #4914
There is no need to fill `entries_removed` when restore. It costs about 30% CPU of the whole restore process.
Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/V3/PageDirectory.h [hunk 1]
```diff
@@ -360,7 +362,7 @@ class PageDirectory
 
     bool tryDumpSnapshot(const ReadLimiterPtr & read_limiter = nullptr, const WriteLimiterPtr & write_limiter = nullptr);
 
-    PageEntriesV3 gcInMemEntries();
+    PageEntriesV3 gcInMemEntries(bool need_remove_blob = true);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/PageDirectory.h:365]:
```suggestion
    // Perform a GC for in-memory entries and return the removed entries.
    // If `return_removed_entries` is false, then just return an empty set.
    PageEntriesV3 gcInMemEntries(bool return_removed_entries = true);
```

File: dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp [hunk 2]
```diff
@@ -44,7 +44,7 @@ PageDirectoryPtr PageDirectoryFactory::createFromReader(String storage_name, WAL
 
     // After restoring from the disk, we need cleanup all invalid entries in memory, or it will
     // try to run GC again on some entries that are already marked as invalid in BlobStore.
-    dir->gcInMemEntries();
+    dir->gcInMemEntries(false);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:47]:
```suggestion
    dir->gcInMemEntries(/*return_removed_entries=*/false);
```
Add comment about why `false` is ok

File: dbms/src/Storages/Page/V3/PageDirectory.h [hunk 3]
```diff
@@ -224,13 +224,15 @@ class VersionedPageEntries
...
+        const PageLock & page_lock,
+        bool need_fill_entries_removed);
     bool derefAndClean(
         UInt64 lowest_seq,
         PageIdV3Internal page_id,
         const PageVersion & deref_ver,
         Int64 deref_count,
-        PageEntriesV3 & entries_removed);
+        PageEntriesV3 & entries_removed,
+        bool need_fill_entries_removed);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/PageDirectory.h:235]: Make the interface to use type `PageEntriesV3 *` for `entries_removed`, and send `nullptr` when we don't need to fill the vector.
This is a unified style with how we pass the param `normal_entries_to_deref`
```
    bool cleanOutdatedEntries(
        UInt64 lowest_seq,
        std::map<PageIdV3Internal, std::pair<PageVersion, Int64>> * normal_entries_to_deref,
        PageEntriesV3 * entries_removed,
        const PageLock & page_lock);
    bool derefAndClean(
        UInt64 lowest_seq,
        PageIdV3Internal page_id,
        const PageVersion & deref_ver,
        Int64 deref_count,
        PageEntriesV3 * entries_removed);
```

**Review by JaySon-Huang — COMMENTED** [review]
Is there any performance result && frame graph for this improvement?

**Comment by hehechen** [comment]
> Is there any performance result && frame graph for this improvement?

A total of 100GB of data, each page is 8kb, the `restore` time is reduced from 15905ms to 11581ms.

File: dbms/src/Storages/Page/V3/PageDirectory.h [hunk 4]
```diff
@@ -360,7 +360,9 @@ class PageDirectory
 
     bool tryDumpSnapshot(const ReadLimiterPtr & read_limiter = nullptr, const WriteLimiterPtr & write_limiter = nullptr);
 
-    PageEntriesV3 gcInMemEntries();
+    // Perform a GC for in-memory entries and return the removed entries.
+    // If `return_removed_entries` is false, then just return an empty set.
+    PageEntriesV3 gcInMemEntries(bool return_removed_entries = true);
```
Comment (jiaqizho) [inline dbms/src/Storages/Page/V3/PageDirectory.h:365]: `return_removed_entries` how about change to `record_removed_entries` ?

**Review by jiaqizho — APPROVED** [review]
LGTM with a comment

**Pull request merged**